src/sema/ast.rs

```diff
@@ -61,6 +61,7 @@ pub enum Type {
     /// There is no way to declare value in Solidity (should there be?)
     Value,
     Void,
+    #[default]
```

Comment:
I don't think `Type` should have a default value. Just does not make any sense.

Reply:
Why do you think that it doesn't make any sense for types to be `Unreachable` by default? I argue the opposite: It makes sense because in tests, I need to work with data structures containing `Type`s, but I don't care about that part. So I want to leave them uninitialized.

We could also pick `Unresolved` by default. We just say that, an uninitiated type ("default") is unresolved or unreachable. To me it makes sense to declare either as the default value. And I'd expect our codegen to refuse CFGs containing `Unreachable` or `Unresolved` types anyways. So what are problems introduced with doing that you're afraid of?

Reply:
Can you point me to code that is made easier by having a default for `Type`?

Reply:
[Sure](https://github.com/hyperledger/solang/pull/1516/files#diff-65527b752e88b0b1678c19e4d555d0db9e9f07c8c60de9050ef8c5e5086ebbdaR476-R499)

And this is just one place, there are more existing tests like that, and more to be written in the future.

Reply:
You're trying to do:
```rust
                 fields: vec![
                     Parameter {
                         ty: Type::Address(false),
                         ..Default::default()
                     },
                     Parameter {
                         ty: Type::Uint(256),
                         ..Default::default()
                     },
                     Parameter {
                         ty: Type::Uint(256),
                         ..Default::default()
                     },
                 ],
```
I think this can be better achieved with a function

```rust
impl Parameter {
    fn default(ty: Type) -> Self {
        Parameter {
            ty,
           indexed: false,
           ...
 
        }
    }
}
```
Now your test code is even shorter:

```rust
                 fields: vec![
                     Parameter::default(Type::Address(false)),
                     Parameter::default(Type::Uint(256)),
                     Parameter::default(Type::Uint(256)),
                 ],
```

Reply:
I think this is a bad idea.

The `Default` trait is used extensively across many rust crates.

Why do you desperately want to work around that?

Reply:
Actually with builders, we could do something close to what you suggested but without having `Type` to implement default, but in a more idiomatic manner.

However I still don't see why `Type` can not implement `Default`

Reply:
Because does not have an natural default. You keep on changing your mind on which one should be default - that's a clear indicator that there is no good default value.

Writing code is about not doing anything surprising/unexpected. I think making some random type the default is pretty much the definition of doing the unexpected. If you had this enum:

It's just like:
```rust
enum Coin {
    Penny,
    Nickel,
    Dime,
    Quarter,
}
```
What default value would you pick?

Reply:
> Because does not have an natural default.

Rusts [Default](https://doc.rust-lang.org/std/default/trait.Default.html) has **nothing** to whether there is a natural default. I quote:

> Sometimes, you want to fall back to some kind of default value, and don’t particularly care what it is.

And this is also _exactly_ why I want to have it implement `Default`.

>  You keep on changing your mind on which one should be default - that's a clear indicator that there is no good default value.
> Writing code is about not doing anything surprising/unexpected. I think making some random type the default is pretty much the definition of doing the unexpected. If you had this enum

It doesn't matter which value we pick by _definition_ of the `Default` trait. But choosing `Unreachable` or `Unresolved` should give us an extra safeguard for bugs in codegen.

I still don't see why this can't implement `Default`

Reply:
The impression I have about setting a default for `Type` is that either there is a default type in Solidity or the compiler has an internal default type. Neither option is true. We must declare a specific type in Solidity and the runtime needs to know the type of a value in order to operate it.

Types must be explicit in a compiler, so I do not like the idea of choosing a default for it. As Sean pointed, there is no default type in Solidity nor in a compiler. We should enforce everyone to explicitly choose one.

Reply:
I also believe `Type` should not have a fallback option, since types should be explicit. When you are executing a program in your CPU, it does not assume a fallback type in case it is not specified. It does not know what to do.

Reply:
Ok, if it's too concerning even if `Unreachable` was picked. I changed it so it's only possible to use that in the tests. And then we can solve the problem outside tests with builders. WDYT

Reply:
> The impression I have about setting a default for Type is that either there is a default type in Solidity or the compiler has an internal default type. Neither option is true.
> We must declare a specific type in Solidity and the runtime needs to know the type of a value in order to operate it.
> I also believe Type should not have a fallback option

This doesn't seem accurate to me. The existence of the `Unresolved` and `Unreachable` `Type` variants directly contradict the entire reasoning.

Reply:
> Ok, if it's too concerning even if `Unreachable` was picked. I changed it so it's only possible to use that in the tests. And then we can solve the problem outside tests with builders. WDYT

This makes the code even worse.

You're probably thinking, who cares what the default is, I just want to be able to `#![derive(Default)]`. I just don't agree with that, some things don't have a default value and should not have one.

Reply:
`Unreachable` has a very specific meaning - control flow should never reach this point
`Unresolved` should never occur in codegen

Reply:
> This makes the code even worse.

Why? It makes writing tests easier, making the code arguably better.

> some things don't have a default value

Again, the idea of the `Default` trait is not to assign a "meaningful" default (for example `bool` and numeric values have a default value too). It's just a way to tell that it's possible to use this type without specifying what the value it is. And we need _exactly_ that in the tests. Now the concern is that implementing this makes it easier to accidentally use the value of `Type::default()` implicitly. Which, by your own reasoning, would be mitigated by picking `Unreachable` or `Unresolved` as the default value. And further giving in, by only implementing it for tests, it's not even possible.

However this is going nowhere, and blocking off the entire contribution. I just tried to make writing unit tests easier using some Rust Language intrinsic. Unfortunately this is being opposed to the point it's starting to get very time consuming is discouraging. I removed the `Default` implementation on type and implemented a helper on `Parameter`, which you liked better, instead